draft-ralston-mimi-linearized-matrix.md

```diff
@@ -687,17 +687,218 @@ The hub server MUST enforce the room version's algorithms upon the room. Partici
+   "signatures": {
+      "@alice:example.org": {
+         "m.mls.v1.credential.ed25519:ABCDEF": "<unpadded base64 signature>"
+      }
+   }
+}
+~~~
+
+`device_id` is the client's device ID ({{int-device-id}}). `user_id` is the user ID ({{int-user-id}})
+to which the device ID belongs. `algorithms` are the encryption algorithms the device supports, and
+SHOULD contain at least `m.mls.v1.dhkemx25519-aes128gcm-sha256-ed25519`.
+
+When a device supports `m.mls.v1.dhkemx25519-aes128gcm-sha256-ed25519`, it MUST specify its basic
+credential with the `m.mls.v1.credential.ed25519` key algorithm.
+
+`keys` is an object containing each algorithm-specific key (or keys) for the device. The fields for
+the object form a key ID, with the device ID representing the "key version", as per {{int-signing}}.
+
+All top-level fields in the object above MUST be supplied.
+
+For each of the device's `keys`, a valid signature MUST be produced. If there is a missing signature
+from any of the keys, or from the `user_id`, the device information is considered invalid. Invalid
+devices MUST NOT be members of the MLS group, and are removed if already members prior to the device
+information becoming invalid.
+
+## Group Creation
+
+After the `m.room.create` event and other initial state events for the room are sent, the room creator
+MUST establish the appropriate MLS group. This is sent as an `m.mls.commit` event ({{int-ev-mls-commit}}).
+Afterwards, the remaining devices are added as normal ({{int-mls-add-remove}}).
+
+Ideally, the `m.room.create` event would also contain the initial public group state, however doing
+so would mean either tracking an independent MLS group ID or allowing the client to specify the room
+ID. While servers MAY allow the client to specify the room ID, servers usually have better context
+for which localparts (see {{int-room-id}}) are already claimed by other rooms. Having independent
+group IDs and room IDs can lead to confusion and a similar sort of namespacing issue (a room creator
+can create a conflicting group ID). Instead, the server (usually) creates the room on behalf of the
+client, allowing the client to then send the initial public group state to the room for other MLS
+members.
+
+## Updating Group State {#int-mls-add-remove}
+
+This document does not provide a way to send proposals to the MLS group, meaning all commits MUST only
+contain proposals which are sent by the same member (see {{Section 12 of I-D.ietf-mls-protocol}}).
+
+All commits are encoded as `m.mls.commit` events ({{int-ev-mls-commit}}) and are sent to the room.
+These commits are additionally encoded using `PublicMessage`, giving servers visibility on the contents
+of the commits. Upon receiving the event (see {{int-receiving-events}}), the hub server MUST additionally
+validate that any membership changes match what is possible with the room membership:
+
+* Devices can only be added to the group if they belong to a user which is joined to the room. It is
```

Comment:
shouldn't invited users be added in rooms that allow them access to room history?  Or is that not a thing in LM?

Reply:
bleh, it is. Will have to figure out how that works exactly.